mcs/class/Mono.Debugger.Soft/Test/dtest.cs

```diff
@@ -4526,5 +4526,122 @@ public void Bug59649 ()
 		AssertValue (3.0, f);
 
 	}
+
+	private void HandleEvents()
+	{
+		string failMessage = null;
+		var alreadyStopped = false;
+		while (true)
+		{
+			var nextEventSet = vm.GetNextEventSet(5000);
+			if (nextEventSet == null)
+				break;
+
+			foreach (var e in nextEventSet.Events)
+			{
+				if (e.EventType == EventType.AssemblyLoad)
+				{
+					var assemblyload = e as AssemblyLoadEvent;
+					var amirror = assemblyload.Assembly;
+					if (amirror.GetName().Name.Contains("dtest-app"))
+					{
+						var methodMirror = amirror.EntryPoint.DeclaringType.GetMethod("attach_break");
+						vm.SetBreakpoint(methodMirror, 0);
+					}
+				}
+
+				if (e.EventType == EventType.TypeLoad)
+				{
+					Assert.Fail("Unexpected TypeLoadEvent");
+				}
+
+				if (e.EventType == EventType.Breakpoint)
+				{
+					var typeLoadRequest = vm.CreateTypeLoadRequest();
+					typeLoadRequest.SourceFileFilter = new[] {"dtest-app.cs"};
+					typeLoadRequest.Enable();
+
+					if (alreadyStopped)
+					{
+						Assert.Fail("Unexpected BreakpointEvent");
+					}
+
+					alreadyStopped = true;
+					continue;
+				}
+
+				try
+				{
+					vm.Resume();
+				}
+				catch (VMNotSuspendedException ex)
+				{
+					failMessage = ex.Message;
+				}
+			}
+		}
+
+		if (failMessage != null)
+			Assert.Fail(failMessage);
+	}
+
+	private void ShouldNotSendUnexpectedTypeLoadEventsAndInvalidSuspendPolicyAfterAttach (int port)
+	{
+		string failMessage = null;
+		try
+		{
+			vm = ConnectToPort(port);
+
+			vm.EnableEvents (EventType.AssemblyLoad, EventType.ThreadStart);
+			var vmstart = GetNextEvent();
+			Assert.AreEqual (EventType.VMStart, vmstart.EventType);
+			try
+			{
+				vm.Resume ();
+			}
+			catch (VMNotSuspendedException e)
+			{
+				failMessage = e.Message;
+			}
+			HandleEvents ();
+		}
+		catch (Exception ex) {
+			failMessage = ex.Message;
+		}
+		finally
+		{
+			vm.Exit (0);
+			vm = null;
+		}
+		if (failMessage != null)
+			Assert.Fail (failMessage);
+	}
+
+	[Test]
+	public void ShouldNotSendUnexpectedTypeLoadEventsAndInvalidSuspendPolicyAfterAttachIfSuspendIsTrue()
+	{
+		vm.Exit (0);
+		var port = GetFreePort();
+
+		// Launch the app using server=y,suspend=y
+		var pi = CreateStartInfo (new string[] { $"--debugger-agent=transport=dt_socket,address=127.0.0.1:{port},server=y,suspend=y", "dtest-app.exe", "attach" });
+		var process = Diag.Process.Start (pi);
+
+		ShouldNotSendUnexpectedTypeLoadEventsAndInvalidSuspendPolicyAfterAttach (port);
+	}
+
+	[Test]
+	[Category("NotWorking")]
+	public void ShouldNotSendUnexpectedTypeLoadEventsAndInvalidSuspendPolicyAfterAttachIfSuspendIsFalse()
+	{
+		vm.Exit (0);
```

Comment:
The idea is that the debuggee collects the events and sends them when the debugger attaches.

Reply:
When we starting HandleEvents in test debugger already attached and stopped at the breakpoint. As I understand after that we shouldn't receive "TypeLoad" or "Breakpoint" event in this time. If we received event after the stop at the breakpoint and call Resume after TypeLoad event we incorrect behaviour.